src/communication/websocket/auth.cpp

```diff
@@ -12,19 +12,45 @@
 #include "communication/websocket/auth.hpp"
 
 #include <string>
+#include "utils/variant_helpers.hpp"
 
 namespace memgraph::communication::websocket {
 
 bool SafeAuth::Authenticate(const std::string &username, const std::string &password) const {
-  return auth_->Lock()->Authenticate(username, password).has_value();
+  user_or_role_ = auth_->Lock()->Authenticate(username, password);
+  return user_or_role_.has_value();
 }
 
-bool SafeAuth::HasUserPermission(const std::string &username, const auth::Permission permission) const {
-  if (const auto user = auth_->ReadLock()->GetUser(username); user) {
-    return user->GetPermissions().Has(permission) == auth::PermissionLevel::GRANT;
+bool SafeAuth::HasPermission(const auth::Permission permission) const {
+  auto locked_auth = auth_->Lock();
+  // Update if cache invalidated
+  if (!locked_auth->UpToDate(auth_epoch_) && user_or_role_) {
+    bool success = true;
+    std::visit(utils::Overloaded{[&](auth::User &user) {
+                                   auto tmp = locked_auth->GetUser(user.username());
+                                   if (!tmp) success = false;
+                                   user = std::move(*tmp);
+                                 },
+                                 [&](auth::Role &role) {
+                                   auto tmp = locked_auth->GetRole(role.rolename());
+                                   if (!tmp) success = false;
+                                   role = std::move(*tmp);
+                                 }},
+               *user_or_role_);
+    // Missing user/role; delete from cache
+    if (!success) user_or_role_.reset();
   }
+  // Check permissions
+  if (user_or_role_) {
+    return std::visit(
+        utils::Overloaded{
+            [&](auth::User &user) { return user.GetPermissions().Has(permission) == auth::PermissionLevel::GRANT; },
+            [&](auth::Role &role) { return role.permissions().Has(permission) == auth::PermissionLevel::GRANT; }},
```

Comment:
Can we unify GetPermissions/permissions then we can use one generic lambda, no need for Overloaded